services/api/src/api/jwt_token.py

```diff
@@ -0,0 +1,161 @@
+        if not isinstance(expected_algorithm, ASYMMETRIC_ALGORITHMS):
+            return key.decode("utf-8")
+        if not is_public_key(key):
+            raise RuntimeError("Failed to parse JWT key: the provided key is a private key")
+        return key.public_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo,
+        ).decode("utf-8")
+    except (jwt.InvalidKeyError, KeyError) as err:
+        raise RuntimeError(f"Failed to parse JWT key: {err.args[0]}") from err
+
+
+def fetch_jwt_public_key(
+    url: str,
+    hf_jwt_algorithm: str,
+    hf_timeout_seconds: Optional[float] = None,
+) -> Any:
+    """fetch the public key to decode the JWT token from the input URL
+
+    See https://huggingface.co/api/keys/jwt
+
+    Args:
+        url (str): the URL to fetch the public key from
+        hf_jwt_algorithm (str): the JWT algorithm to use.
+        hf_timeout_seconds (float|None): the timeout in seconds for the external authentication service. It
+            is used both for the connection timeout and the read timeout. If None, the request never timeouts.
+
+    Returns:
+        Any: the public key
+    """
+    try:
+        response = requests.get(url, timeout=hf_timeout_seconds)
+        response.raise_for_status()
+        return parse_jwt_public_key(keys=response.json(), hf_jwt_algorithm=hf_jwt_algorithm)
+    except Exception as err:
+        raise JWKError(f"Failed to fetch or parse the JWT public key from {url}. ", cause=err) from err
+
+
+sub_schema = {
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "type": "object",
+    "properties": {
+        "repoName": {"type": "string"},
+        "repoType": {"type": "string", "enum": ["dataset"]},
+        "read": {"type": "boolean", "enum": [True]},
+    },
+    "required": ["repoName", "repoType", "read"],
+}
+
+
+def is_jwt_valid(dataset: str, token: Any, public_key: Optional[Any], algorithm: Optional[str]) -> bool:
+    """
+    Check if the JWT is valid for the dataset.
+
+    The JWT is decoded with the public key, and the "sub" claim must be:
+      {"repoName": <...>, "repoType": "dataset", "read": true}
+    where <...> is the dataset identifier.
+
+    Returns True only if all the conditions are met. Else, it returns False.
+
+    Args:
+        dataset (str): the dataset identifier
+        token (Any): the JWT token to decode
+        public_key (Any|None): the public key to use to decode the JWT token
+        algorithm (str|None): the algorithm to use to decode the JWT token
+
+    Returns:
+        bool: True if the JWT is valid for the input dataset, else False
+    """
+    if not public_key or not algorithm:
+        return False
+    try:
+        decoded = jwt.decode(jwt=token, key=public_key, algorithms=[algorithm], options={"require": ["exp", "sub"]})
+    except Exception as err:
+        print(err)
+        return False
+    sub = decoded.get("sub")
```

Comment:
If I understand correctly, `sub` is expected to be a JSON encoded object, correct?
If true, note that `sub` claims are strings as per the JWT standard.
WDYT about passing this info in the JWT payload directly?

Ref: https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.2

Reply:
oh yes I agree

Reply:
Maybe make `sub` `/dataset/user/repo` :thinking: (and extra info in the payload)

Reply:
> No need for `"repoType": "dataset"`, right?

yes

Reply:
Done